tests/phpunit/tests/theme/wpThemeJson.php

```diff
@@ -1483,6 +1483,100 @@ public function test_merge_incoming_data_color_presets_with_same_slugs_as_defaul
 		$this->assertEqualSetsWithIndex( $expected, $actual );
 	}
 
+	/**
+	 * @ticket 54640
+	 */
+	public function test_merge_incoming_data_presets_use_default_names() {
+		$defaults   = new WP_Theme_JSON(
+			array(
+				'version'  => WP_Theme_JSON::LATEST_SCHEMA,
+				'settings' => array(
+					'typography' => array(
+						'fontSizes' => array(
+							array(
+								'name' => 'Small',
+								'slug' => 'small',
+								'size' => '12px',
+							),
+							array(
+								'name' => 'Large',
+								'slug' => 'large',
+								'size' => '20px',
+							),
+						),
+					),
+				),
+			),
+			'default'
+		);
+		$theme_json = new WP_Theme_JSON(
+			array(
+				'version'  => WP_Theme_JSON::LATEST_SCHEMA,
+				'settings' => array(
+					'typography' => array(
+						'fontSizes' => array(
+							array(
+								'slug' => 'small',
+								'size' => '1.1rem',
+							),
+							array(
+								'slug' => 'large',
+								'size' => '1.75rem',
+							),
+							array(
+								'name' => 'Huge',
+								'slug' => 'huge',
+								'size' => '3rem',
+							),
+						),
+					),
+				),
+			),
+			'theme'
+		);
+		$expected   = array(
+			'version'  => WP_Theme_JSON::LATEST_SCHEMA,
+			'settings' => array(
+				'typography' => array(
+					'fontSizes' => array(
+						'theme'   => array(
+							array(
+								'name' => 'Small',
+								'slug' => 'small',
+								'size' => '1.1rem',
+							),
+							array(
+								'name' => 'Large',
+								'slug' => 'large',
+								'size' => '1.75rem',
+							),
+							array(
+								'name' => 'Huge',
+								'slug' => 'huge',
+								'size' => '3rem',
+							),
+						),
+						'default' => array(
+							array(
+								'name' => 'Small',
+								'slug' => 'small',
+								'size' => '12px',
+							),
+							array(
+								'name' => 'Large',
+								'slug' => 'large',
+								'size' => '20px',
+							),
+						),
+					),
+				),
+			),
+		);
+		$defaults->merge( $theme_json );
+		$actual = $defaults->get_raw_data();
+		$this->assertEqualSetsWithIndex( $expected, $actual );
```

Comment:
For `settings.typography.fontSizes.theme` elements that contain an existing `name` key/value pair, these maintain the order in which they are supplied.

i.e.
```php
$input = array(
    'name' => 'Huge',
    'slug' => 'huge',
    'size' => '3rem',
)

// WP_Theme_JSON->merge().

$output = array(
    'name' => 'Huge',
    'slug' => 'huge',
    'size' => '3rem',
);
```

However, if the element is missing a `name` key/value pair and this is added by the call to `WP_Theme_JSON->merge()`, then the name key is added to the end of the array.

i.e.
```php
$input = array(
    'slug' => 'huge',
    'size' => '3rem',
)

// WP_Theme_JSON->merge().

$output = array(
    'slug' => 'huge',
    'size' => '3rem',
    'name' => 'Huge',
);
```

Is this the expected behaviour, or should the keys be in a consistent order?

Core is implementing stricter testing (i.e. `assertSameSetsWithIndex()`).

For this to be implemented, either:
- if the behaviour above is expected, the `$expected` value should change to reflect the expected behaviour that the keys are in the order they are added to the array. OR
- if the behaviour above isn't expected, the source needs to be modified to sort the keys into a consistent order.

The whole test class needs to be revisited for stricter testing, so I fully understand why the `assertEqualSetsWithIndex()` made sense to use for consistency. The action taken in this instance will help inform how we can proceed with updating the other assertions in test class.

Reply:
>Is this the expected behaviour, or should the keys be in a consistent order?

In this specific case at least (although I'd expect to most if not all other cases) the order of keys is completely irrelevant. Having said that, do you think it gives much value to reflect this change in the `expected`?

Reply:
Thanks for clarifying!

In terms of giving value to the behaviour being tested in this test method, if the order of keys were ever deemed important, then it would add value. However, as you said, that's not the case in this instance.

In terms of testing/documentation/discussion, we should use the strictest possible assertion to maximize the stability of source and our awareness of what's changing. For example, if something changes in source, stricter assertions (therefore, test failures) can reveal side effects that we may not have been aware of. These side effects can then be evaluated and the result will either be changes to source or changes to unit tests.

Reply:
I updated to use `assertSameSetsWithIndex`.